connectors/sources/outlook.py

```diff
@@ -0,0 +1,944 @@
+        source_datetime, date
+    ):
+        return source_datetime.strftime("%Y-%m-%d")
+    else:
+        return source_datetime
+
+
+class TokenFetchFailed(Exception):
+    """Exception class to notify that connector was unable to fetch authentication token from Graph API"""
+
+    pass
+
+
+class UsersFetchFailed(Exception):
+    """Exception class to notify that connector was unable to fetch users from Active Directory"""
+
+    pass
+
+
+class UnauthorizedException(Exception):
+    """Exception class unauthorized calls"""
+
+    pass
+
+
+class Forbidden(Exception):
+    pass
+
+
+class NotFound(Exception):
+    pass
+
+
+class SSLFailed(Exception):
+    pass
+
+
+class RootCAAdapter(requests.adapters.HTTPAdapter):
+    """Class to verify SSL Certificate for Exchange Servers"""
+
+    def __init__(self):
+        self.ssl_verify = False
+        self.ssl_ctx = None
+
+    def cert_verify(self, conn, url, ssl_certificate_file, cert):
+        try:
+            super().cert_verify(
+                conn=conn, url=url, verify=self.ssl_verify, cert=self.ssl_ctx
+            )
+        except Exception as exception:
+            raise SSLFailed(
+                f"Something went wrong while verifying SSL certificate. Error: {exception}"
+            ) from exception
+
+
+class ExchangeUsers:
+    """Fetch users from Exchange Active Directory"""
+
+    def __init__(self, ad_server, domain, exchange_server, user, password, ssl_enabled):
+        self.ad_server = Server(host=ad_server)
+        self.domain = domain
+        self.exchange_server = exchange_server
+        self.user = user
+        self.password = password
+        self.ssl_enabled = ssl_enabled
+
+    @cached_property
+    def _create_connection(self):
+        return Connection(
+            server=self.ad_server,
+            user=self.user,
+            password=self.password,
+            client_strategy=SAFE_SYNC,
+            auto_bind=True,  # pyright: ignore
+        )
+
+    async def get_users(self):
+        warnings.filterwarnings("default")
+        ldap_domain_name_list = ["DC=" + domain for domain in self.domain.split(".")]
+        search_query = ",".join(map(str, ldap_domain_name_list))
+
+        has_value_for_normal_users, _, response, _ = self._create_connection.search(
+            search_query,
+            SEARCH_FILTER_FOR_NORMAL_USERS,
+            attributes=["mail"],
+        )
+
+        (
+            has_value_for_admin_users,
+            _,
+            response_for_admin,
+            _,
+        ) = self._create_connection.search(
+            search_query,
+            SEARCH_FILTER_FOR_ADMIN,
+            attributes=["mail"],
+        )
+
+        if not (has_value_for_normal_users or has_value_for_admin_users):
+            raise UsersFetchFailed(
+                "Error while fetching users from Exchange Active Directory."
+            )
+
+        response.extend(response_for_admin)
```

Comment:
I didn't check the documentation, just want to make sure:

1. The `search` operation will not raise exceptions but always return something in the first return value (e.g. `has_value_for_normal_users`) to indicate the search result.
2. The code proceeds as long as either of  `has_value_for_normal_users` and `has_value_for_admin_users` is `True`. Make sure `response` is not `None` if `has_value_for_normal_users` is `False`.

Reply:
> The search operation will not raise exceptions but always return something in the first return value (e.g. has_value_for_normal_users) to indicate the search result.

Yes, it will not raise an exception instead return a status to indicate the result

> The code proceeds as long as either of has_value_for_normal_users and has_value_for_admin_users is True. Make sure response is not None if has_value_for_normal_users is False.

Sure